[WFLY-19792] MP multiple app support

kabir opened the pull request:
https://issues.redhat.com/browse/WFLY-19792

jmesnil commented on microprofile/design-doc-template.adoc:
```diff
@@ -0,0 +1,178 @@
+# - management
+# if missing, add it to _data/wildfly-categories and use its id
+categories:
+- MicroProfile
```
s/MicroProfile/microprofile

jmesnil commented on microprofile/design-doc-template.adoc:
```diff
@@ -0,0 +1,178 @@
+---
```
@kabir Could you rename the file `design-doc-template` before making the PR ready for review?

fabiobrz commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+
+=== Deployments
+
+There are no incompatible changes to the behaviour deployments.
```
s/behaviour deployments/behaviour of deployments/ ?

fabiobrz commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+
+== Implementation Plan
+
+// __<This section is optional. If you have a complex feature which can not be delivered all in one go, suggest the strategy.>__
```
Can we remove template comments? Here and below.

fabiobrz commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+* MicroProfile Reactive Messaging - it will point out that only .war deployments are supported, and that although it might work in other deployment types, the behaviour is undefined.
+* MicroProfile Rest Client - it will point out that only .war deployments are supported, and that although it might work in other deployment types, the behaviour is undefined.
+* OpenTelemetry / MicroProfile Telemetry - the documentation will be updated to detail that the deployment scenario is supported and what to expect (especially with regard to duplicate names and their tags)
+* Micrometer - the documentation will be updated to detail that the deployment scenario is supported and what to expect
```
s/deployment scenario/multi deployment scenario/ ?

fabiobrz commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+ developer: kabir
+ sme:
+  - jmesnil
+ outside-perspective:
```
Should this be set to me?

fabiobrz commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+** Each .war contributes its healthiness checks to the global server healthiness.
+** When multiple .war archives are deployed, if at least *one* defines the `mp.health.disable-default-procedures` config to `true`, it affects the whole server configuration (as described in https://docs.wildfly.org/36/Admin_Guide.html#disabling-default-server-procedure[("Disabling Default Server Procedures" documentation)]). This multi-deployment behaviour is not specified in the MicroProfile He ...
+* MicroProfile OpenAPI
+** MP OpenAPI subsystem must install an OpenAPI endpoint and model per host (rather than per deployment) and allow applications to contribute components/paths/etc. during deployment.
```
BTW - I am just wondering whether - being OpenAPI the spec which needs more work - we could add a link to the community documentation which is being draft in https://github.com/wildfly/wildfly/pull/19003/files. WDYT?

pferraro replied:
`host` does not refer to the OpenAPI model, but rather refers to an Undertow host (i.e. a virtual host).  I don't think WildFly uses this term in any other context.

fabiobrz reviewed (changes requested):
Thanks for this proposal @kabir. 
I added some comments to clarify things a bit, feel free to let me know what you think.

pferraro commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+** When multiple .war archives are deployed, if at least *one* defines the `mp.health.disable-default-procedures` config to `true`, it affects the whole server configuration (as described in https://docs.wildfly.org/36/Admin_Guide.html#disabling-default-server-procedure[("Disabling Default Server Procedures" documentation)]). This multi-deployment behaviour is not specified in the MicroProfile He ...
+* MicroProfile OpenAPI
+** MP OpenAPI subsystem must install an OpenAPI endpoint and model per host (rather than per deployment) and allow applications to contribute components/paths/etc. during deployment.
+** "Global" properties of the OpenAPI model that are not concordant between applications will be provided by the following host-specific MP Config properties:
```
MP Config properties are not validated - so any property _could_ be defined this way, but only the ones listed here will be read by the container.

fabiobrz replied:
Which - IIUC - means that:

> just `externalDocs` and `info` OpenAPI document properties can be defined this way

Correct?
If not please provide some better explanation in the analysis document, which should give the reader - not necessarily a developer - a quick and easy preview of what is going to happen.

pferraro replied:
Perhaps I did not understand your question.  Can you rephrase?

kabir commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+
+=== User Stories
+
+As an example the user may want to deploy two applications using MicroProfile Config. Each application should only see their own config, and config specified at a 'higher' level such as via environment variables and system properties. For this case, each application should behave as if it is the only application deployed on the server.
```
@pferraro @jasondlee Do you have some examples for your integrations?

Feel free to just leave comments here with suggested text, and I can incorporate those into the proposal.

pferraro replied:
WRT MP OpenAPI: Consumers of an OpenAPI document should assume that it describes all REST services accessible from associated virtual host.  Whether these services are provided by one application or many should be opaque to the user.  From the perspective of the application developer, an application should be able to define its OpenAPI model (via the mechanisms described in the spec) independently from any other applications that might be associated with the same virtual host.

kabir commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+* link:#affected-projects-or-components[Affected Components] - if changes are needed to the code of one of the components, that component will be mentioned here.
+* link:#requirements[Requirements] - components listed here are covered by this RFE, along with a brief description of the expected behavious. if there are some corner cases, or something working  differently from how this would be naively interpreted, these will be mentioned here for the affected components. This may also include the link:#non-requirements[Non-Requirements] sub-section.
+* link:#related-issues[Related Issues] - for each MicroProfile specification included in *this* umbrella RFE, the issue tracking the work is listed here. Typically, these are issues with an 'incorporates' link from https://issues.redhat.com/browse/WFLY-19792. If any separate RFEs are needed those will be listed here.
+* link:#affected-projects-or-components[Affected Components] - if changes are needed in other components, the changes are summarised here.
```
Removed

kabir commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+** Further each deployment may define its own config properties via a `META-INF/microprofile-config.properties`, or custom `ConfigSource` and `ConfigSourceProvider` implementations. These own config properties are local to the deployment, and must only be visible from within this deployment. So if we have A.war and B.war, A.war should not see properties local to B.war and vice versa. Of course, A ...
+* MicroProfile Fault Tolerance
+** Already tested with multiple .war deployment with metrics providers; expanding of the test case can be considered.
+** Each deployment behaves as if it were deployed on its own, however, metrics contribute to the same metric collector.
```
Not sure, let's see what @rhusar has to say

kabir commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+* MicroProfile Fault Tolerance
+** Already tested with multiple .war deployment with metrics providers; expanding of the test case can be considered.
+** Each deployment behaves as if it were deployed on its own, however, metrics contribute to the same metric collector.
+* MicroProfile Reactive Messaging - This already has tests for multiple .war deployment, done as part of https://issues.redhat.com/browse/WFLY-19759[WFLY-19759] a while ago.
```
I don't think there is more than the MP Config aspect. The CDI beans containing the @Incoming / @Outgoing annotations etc are local to the deployment. But I think that is probably something we can take to be inferred from how CDI works?

kabir commented:
Hi @fabiobrz I have addressed the comments I was able to, and have asked around a bit for the remaining points. Hopefully we will get some answers soon

pferraro commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+
+For these straight-forward cases, we will populate the following sections:
+
+* link:#affected-projects-or-components[Affected Components] - if changes are needed to the code of one of the components, that component will be mentioned here.
```
No smallrye-openapi changes are required for this.

jasondlee commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,221 @@
+is not much that needs to be done to ensure this works.
+** This also requires MicroProfile Config. No changes in the implementation are needed to have this work with multiple deployments.
+* MicroProfile Telemetry / OpenTelemetry
+** There are no explicit tests for multiple war deployments, but the context propagation test does deploy two separate OpenTelemetry applications to test the propagation functionality, so multiple apps are tested implicitly
```
This test deploys two otel applications: https://github.com/wildfly/wildfly/blob/main/testsuite/integration/expansion/src/test/java/org/wildfly/test/integration/observability/opentelemetry/ContextPropagationTestCase.java#L44-L52

kabir replied:
Adding link to proposal

fabiobrz reviewed (changes requested):
@kabir I've dropped some minor, and hopefully final, comments. 
Feel free to let me know what you think.

rhusar commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,224 @@
+== Admin Clients
+
+No admin client issues.
+== Security Considerations
```
```suggestion

== Security Considerations
```

rhusar commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,224 @@
+
+== Community Documentation
+
+// __<Describe how this feature will be documented or illustrated. Generally a feature should have documentation as part of the PR to wildfly main, or as a follow up PR if the feature is in wildfly-core. In some cases though the feature will bring additional content (such as quickstarts, guides, etc.). Indicate which of these will happen>__
```
```suggestion
```

rhusar commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,224 @@
+categories:
+- microprofile
+stability-level: default
+# Specify the Feature Development tracker issue for the feature.
+# This must be an issue tracked in https://github.com/orgs/wildfly/projects/7/views/1.
+# To create a Feature Development tracker issue, go to  https://github.com/wildfly/wildfly-proposals/issues/new/choose
+# and select 'Feature Development'
```
```suggestion
```

rhusar commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,224 @@
+issue: https://github.com/wildfly/wildfly-proposals/issues/731
+feature-team:
+ developer: kabir
+ sme:
+  - jmesnil
```
Should this field enumerate all feature participants since the developer field is not a list but a single item? I find it quote odd that the actual participating SMEs are not listed anywhere?

```suggestion
 sme:
  - jmesnil
  - jamezp
  - jasondlee
  - pferraro
  - rhusar
```

kabir commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,224 @@
+* MicroProfile OpenAPI - update documentation to describe how to specify global OpenAPI model properties when multiple applications are deployed
+* MicroProfile Reactive Messaging - it will point out that only .war deployments are supported, and that although it might work in other deployment types, the behaviour is undefined.
+* MicroProfile Rest Client - it will point out that only .war deployments are supported, and that although it might work in other deployment types, the behaviour is undefined.
+* OpenTelemetry / MicroProfile Telemetry - the documentation will be updated to detail that the deployment scenario is supported and what to expect (especially with regard to duplicate names and their tags)
```
@jasondlee have you done this already? If not please file a Jira, link to https://issues.redhat.com/browse/WFLY-19792 and update this section.

jasondlee replied:
@kabir https://issues.redhat.com/browse/WFLY-21006

kabir commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,224 @@
+* MicroProfile Reactive Messaging - it will point out that only .war deployments are supported, and that although it might work in other deployment types, the behaviour is undefined.
+* MicroProfile Rest Client - it will point out that only .war deployments are supported, and that although it might work in other deployment types, the behaviour is undefined.
+* OpenTelemetry / MicroProfile Telemetry - the documentation will be updated to detail that the deployment scenario is supported and what to expect (especially with regard to duplicate names and their tags)
+* Micrometer - the documentation will be updated to detail that the multi-deployment scenario is supported and what to expect when war applications are deployed in parallel
```
@jasondlee have you done this already? If not please file a Jira, link to https://issues.redhat.com/browse/WFLY-19792 and update this section.

jasondlee replied:
Same as above: https://issues.redhat.com/browse/WFLY-21006

fabiobrz replied:
@jasondlee cpuld you update the relevant section, please?

kabir replied:
@fabiobrz added to line 72

fabiobrz replied:
Sorry for the typo, @jasondlee, s/cpuld/could/, see my previous question:

> if not, please file a Jira, link to https://issues.redhat.com/browse/WFLY-19792 and _update this section_.

So well... it was _just_ about filing a new Jira issue, and have it mentioned it in this section.

Thanks @kabir for doing that.

fabiobrz commented on microprofile/WFLY-19792-mp-multiple-wars.adoc:
```diff
@@ -0,0 +1,224 @@
+* MicroProfile OpenAPI - update documentation to describe how to specify global OpenAPI model properties when multiple applications are deployed
+* MicroProfile Reactive Messaging - it will point out that only .war deployments are supported, and that although it might work in other deployment types, the behaviour is undefined.
+* MicroProfile Rest Client - it will point out that only .war deployments are supported, and that although it might work in other deployment types, the behaviour is undefined.
+* OpenTelemetry / MicroProfile Telemetry - the documentation will be updated to detail that the deployment scenario is supported and what to expect (especially with regard to duplicate names and their tags) ()
```
@kabir I think a leftover or minor typo, i.e. empty ` ()` here.

rhusar reviewed (approved):
LGTM
